### PR TITLE
expand scope of rethrown gradle errors

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Collect the HTTP exceptions into a List, add the exclusive access and zip exception cases that can also occur when downloading.

Add a permission exception check.

Add a exited abnormally check which prints the full message.

### CODE REVIEW

**Comment by zanderso** [comment]
Whoops, I didn't know you were already looking at this and I think @blasten was taking a look as well. Sorry for the confusion!

**Comment by jonahwilliams** [comment]
Ahh, no problem if someone else wants to take it over. I was just digging through the crash logs today and a lot of these stood out.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 1]
```diff
@@ -304,6 +304,15 @@ String _locateGradlewExecutable(Directory directory) {
   return null;
 }
 
+// Gradle crashes for several known reasons when downloading that are not
+// actionable by flutter.
+const List<String> _kKnownErrorPrefixes = <String>[
+  'java.io.FileNotFoundException: https://downloads.gradle.org',
+  'java.io.IOException: Unable to tunnel through proxy',
+  'java.lang.RuntimeException: Timeout of 120000',
```
Comment (blasten) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Maybe we remove `120000`?

File: packages/flutter_tools/test/general.shard/android/gradle_test.dart [hunk 2]
```diff
@@ -969,6 +969,85 @@ at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)''';
...
+      await expectLater(() async {
+        await checkGradleDependencies();
+      }, throwsToolExit(message: 'does not have permission to execute by your user'));
+    }, overrides: <Type, Generator>{
+      Cache: () => Cache(rootOverride: tempDir),
+      FileSystem: () => fs,
+      ProcessManager: () => mockProcessManager,
+    });
+
+    testUsingContext('throws toolExit if gradle times out waiting for exclusive access to zip', () async {
```
Comment (blasten) [inline packages/flutter_tools/test/general.shard/android/gradle_test.dart]: If possible, would you mind also testing the other error prefixes?

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 3]
```diff
@@ -333,10 +342,21 @@ Future<String> _initializeGradle(FlutterProject project) async {
     );
   } on ProcessException catch (e) {
     final String error = e.toString();
-    if (error.contains('java.io.FileNotFoundException: https://downloads.gradle.org') ||
-        error.contains('java.io.IOException: Unable to tunnel through proxy')) {
+    if (_kKnownErrorPrefixes.any((String candidate) => error.contains(candidate))) {
       throwToolExit('$gradle threw an error while trying to update itself.\n$e');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Maybe advise to just try again. That's fixed this for me in the past. We might also add a TODO here to just do the retry automatically.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 4]
```diff
@@ -333,10 +342,21 @@ Future<String> _initializeGradle(FlutterProject project) async {
...
+    if (error.contains('Permission denied')) {
+      throwToolExit(
+        '$gradle does not have permission to execute by your user.\n'
+        'You should change the ownership of these directories to your user, or '
+        'move the project to a directory with execute permissions.'
+      );
+    }
+    // No idea what went wrong but we can't do anything about it.
+    if (error.contains('ProcessException: Process exited abnormally')) {
+      throwToolExit('$gradle exited abnormally.\n$error');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Maybe advise to run with -v if that will give output from gradle that might be helpful (?). Pointers to the relevant files under user control that might be broken could also be helpful.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 5]
```diff
@@ -333,10 +342,21 @@ Future<String> _initializeGradle(FlutterProject project) async {
...
-    if (error.contains('java.io.FileNotFoundException: https://downloads.gradle.org') ||
-        error.contains('java.io.IOException: Unable to tunnel through proxy')) {
+    if (_kKnownErrorPrefixes.any((String candidate) => error.contains(candidate))) {
       throwToolExit('$gradle threw an error while trying to update itself.\n$e');
     }
+    // gradlew is missing execute.
+    if (error.contains('Permission denied')) {
+      throwToolExit(
+        '$gradle does not have permission to execute by your user.\n'
+        'You should change the ownership of these directories to your user, or '
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Can we say anything more specific than 'these directories' ?

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Done

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Done

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Specified "project directory", but we do show the path to the gradle above. I'll update it to include the original error.

File: packages/flutter_tools/test/general.shard/android/gradle_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/gradle_test.dart]: Done

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Done

**Review by zanderso — APPROVED** [review]

**Review by blasten — APPROVED** [review]
Thanks!

**Pull request merged**